Finance: consistent payment creation

sohkai opened the pull request:
Splits up `newPayment()` into:

- `newScheduledPayment()`: _always_ saves a `RecurringPayment`, for repeating or scheduled future payments
- `newImmediatePayment()`: _only_ for single, immediately-payable transactions

It also modifies the execution behaviour of payments. Previously, it was possible to call `executePayment()` or `receiverExecutePayment()` even if no payments were made (only a single `PaymentFailure` event would be emitted). Now calling these two functions will cause a revert, with more obvious error messages (either you hadn't waited long enough, or there wasn't enough balance in the Vault).

We still allow `newScheduledPayment()` to create a "failing" payment, as I think it'd still be useful to schedule payments even if you can't immediately pay them out.

bingen commented on apps/finance/contracts/Finance.sol:
```diff
@@ -322,9 +345,14 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
+                revert(ERROR_EXECUTE_PAYMENT_TIME);
+            } else {
+                revert(ERROR_EXECUTE_PAYMENT_NUM);
+            }
```
This could be written in 2 lines, as:
```
require(_nextPaymentTime(_paymentId) <= getTimestamp64(), ERROR_EXECUTE_PAYMENT_TIME);
revert(ERROR_EXECUTE_PAYMENT_NUM);
```
Not saying it is better this way, it just came up to my mind and found it kind of nice, but maybe it is clearer as it is. :wink:

sohkai replied:
Another alternative is to use a ternary, but some quick checking with this on remix suggests the ternary is more expensive than the if/else 🤷‍♂️.

bingen commented on apps/finance/test/finance.js:
```diff
@@ -13,6 +14,11 @@ const TokenReturnMissingMock = artifacts.require('TokenReturnMissingMock')
+const assertPaymentFailure = receipt => {
+    const filteredLogs = receipt.logs.filter(log => log.event == 'PaymentFailure')
+    assert.equal(filteredLogs.length, 1, 'should have logged payment failure')
+}
```
Couldn't we use `assertEvent` here?

bingen commented on apps/finance/contracts/Finance.sol:
```diff
@@ -322,9 +345,14 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
+            } else {
+                revert(ERROR_EXECUTE_PAYMENT_NUM);
+            }
+        }
```
Now I see that error messages are slightly different, but I don't think that is necessary, as you usually know which function are you calling, right? (and those functions are both `external`)

sohkai replied:
I think this is reasonable; the function signature should be easy to spot.

izqui commented on apps/finance/contracts/Finance.sol:
```diff
@@ -258,11 +256,36 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
+    */
+   function newPaymentTransaction(address _token, address _receiver, uint256 _amount, string _reference)
+        external
+        authP(CREATE_PAYMENTS_ROLE, arr(_token, _receiver, _amount, MAX_UINT, 1))
```
I think most users will parametrize it with a `>`, as setting a payment with a very low interval can drain the vault.

bingen replied:
Yes, that makes sense, but actually both could be set using an `AND` operator, and then this function would not work.

sohkai replied:
I chose `MAX_UINT64` against `0` since most people will probably want `>`.

I'll document it, but another thing we could do is use `MAX_UINT256` instead here, as a sentinel interval value for "never".

izqui replied:
That makes sense

izqui commented on apps/finance/contracts/Finance.sol:
```diff
@@ -566,15 +600,15 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
     }
 
-    function _executePayment(uint256 _paymentId) internal {
+    function _executePayment(uint256 _paymentId) internal returns (uint256) {
```
We could add an extra return value that is true if the `PaymentFailure` event was emitted, which would remove the need to perform a redundant `_nextPaymentTime(_paymentId) > getTimestamp64()` to decide the revert reason.

sohkai replied:
Was looking at this and I don't think this covers the right cases. `PaymentFailure` is only emitted if at least one payment was due, but the funds weren't available, and needs the `_nextPaymentTime(_paymentId) > getTimestamp64()` in each iteration.

In the check, we want to detect if the payment wasn't due at all for deciding the revert reason.

izqui replied:
Might not be getting the full picture and this is a very small optimization, but the `PaymentFailure` event can only be emitted if there was a payment due that couldn't be performed. If we returned one extra value (`bool lastPaymentFailed`) then in `_executePaymentAtLeastOnce` we could just decide on the revert reason depending on `lastPaymentFailed` rather than re-checking if there was a payment due.

bingen replied:
Yes, I don't get it either. Assuming `paid == 0` back in `_executePaymentAtLeastOnce`, if `_nextPaymentTime(_paymentId) > getTimestamp64()` then it means that it didn't get into the loop of `_executePayment`, so `PaymentFailure` wasn't emitted; while if the opposite holds true, then it means it got into the loop and therefore it went through the `break` and so the event was emitted, right? Maybe I'm missing something too.

sohkai replied:
Ohh yes, I see what @izqui is saying now. Yes, if `paid == 0` and a `PaymentFailure()` event is emitted then we will always be throwing `ERROR_EXECUTE_PAYMENT_NUM`.

I've reordered the `if`s in `_executePaymentAtLeastOnce()` to make this a bit more clear, but decided against the extra return value as it'd make successful calls more expensive.

izqui commented on apps/finance/test/finance.js:
```diff
@@ -1,3 +1,4 @@
+const assertEvent = require('@aragon/test-helpers/assertEvent')
```
I think we should check revert reasons in this contract to ensure that things are reverting with the error that we expected. See https://github.com/aragonone/aragon-court/blob/6d0b322becfb5ff528437c14087ebd07cc918520/test/helpers/assert-revert.js which uses Truffle 4

sohkai commented on apps/finance/contracts/Finance.sol:
```diff
@@ -228,24 +233,17 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
         returns (uint256 paymentId)
```
Good thing nobody's using the params!

izqui replied:
Even if we think so, we shouldn't change the param orders as it is a breaking change.

sohkai commented on apps/finance/contracts/Finance.sol:
```diff
@@ -258,11 +256,36 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
-            _executePayment(paymentId);
-        }
+        // We skip checking how many times the new payment was executed to allow creating new
+        // recurring payments before having enough vault balance
```
`executePayment()` will revert if you try to execute a payment that results in no actual transactions.

Here, in `newPayment()`, I'd like to explicitly allow this, so you can "store" a payment that should be immediately payable (but you can't pay yet, perhaps due to a lack of funds).

The intention is that it's kind of like signing up for a recurring service fee even if you can't pay the first bill when you initially set up the recurring payment.

sohkai commented:
@izqui @bingen Ready for a re-review.

Changes:

- 1fd6ade: modifies some radspec strings, given the scope of the new functions
- 7d00e37: Moved the execution error checking to a new internal function, `_executePaymentAtLeastOnce()`
- 0748ed3: Renamed `newPaymentTransaction()` to `newImmediatePayment()`
- e302575: Renamed `newPayment()` to `newScheduledPayment()` (see https://github.com/aragon/aragon-apps/pull/735 for another renaming for the structs given this change)

sohkai commented on apps/finance/contracts/Finance.sol:
```diff
@@ -781,6 +820,18 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
 
+    // Syntax sugar
+
+    function _arr(address _a, address _b, uint256 _c, uint256 _d, uint256 _e, uint256 _f) internal pure returns (uint256[] r) {
```
This is a bit awkward; for some reason Solidity won't let me use just the `arr()` name :(.

Don't really think we need to port this over back to aragonOS; this is a really specific helper.

bingen replied:
Yes, it looks like. Although that seems to me like a hard question (which ones to include in aragonOS helpers), the possible combinations of parameters are endless.

izqui commented on apps/finance/contracts/Finance.sol:
```diff
@@ -228,28 +234,22 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
     )
         external
-        authP(CREATE_PAYMENTS_ROLE, arr(_token, _receiver, _amount, _interval, _maxRepeats))
+        authP(CREATE_PAYMENTS_ROLE, _arr(_token, _receiver, _amount, uint256(_initialPaymentTime), uint256(_interval), uint256(_maxRepeats)))
```
I think new params should always be added at the end.
